cpp/src/dictionary/add_keys.cu

```diff
@@ -91,16 +93,24 @@ std::unique_ptr<column> add_keys(
                                             mr,
                                             stream)
                          ->release();
-  // The output of lower_bound is INT32 but we need to convert to UINT32.
-  // There are no negative values and uint32 is the same width as int32
-  // so we can create a UINT32 column using the same data.
-  auto const indices_size = static_cast<size_type>(table_indices.front()->size());
-  auto contents           = table_indices.front()->release();
-  auto indices_column     = std::make_unique<column>(data_type{type_id::UINT32},
-                                                 indices_size,
-                                                 std::move(*(contents.data.release())),
-                                                 rmm::device_buffer{0, stream, mr},
-                                                 0);
+  // The output of lower_bound is INT32 but we need to convert to unsigned indices.
+  auto const indices_type = get_indices_type_for_size(keys_column->size());
+  auto indices_column     = [&] {
+    column_view gather_result = table_indices.front()->view();
+    auto const indices_size   = gather_result.size();
+    // we can just use the lower-bound/gather data directly for UINT32 case
+    if (indices_type.id() == type_id::UINT32) {
+      auto contents = table_indices.front()->release();
+      return std::make_unique<column>(data_type{type_id::UINT32},
+                                      indices_size,
+                                      std::move(*(contents.data.release())),
+                                      rmm::device_buffer{0, stream, mr},
+                                      0);
+    }
+    // otherwise we need to convert the gather result
+    column_view cast_view(gather_result.type(), indices_size, gather_result.head(), nullptr, 0);
```

Comment:
Is `cast_view` different from `gather_result`?

Reply:
`gather_result` may have a null mask where `cast_view` will not

Reply:
If `gather_result` has nulls, won't those elements be invalid after the cast?

Reply:
Actually, the `gather_results` will not have any nulls but will be `nullable` and this will fail some gtests. Using the `cast_view` here is just an easy way to create the output indices column without any nulls and also not be `nullable`.